HDDS-5296. Replication Manager should not sent replica command to DN which is not in IN_SERVICE opState

ChenSammi opened the pull request:
https://issues.apache.org/jira/browse/HDDS-5296

sodonnel commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackAware.java:
```diff
@@ -275,28 +277,32 @@ private Node chooseNode(List<Node> excludedNodes, Node affinityNode,
+      if (nodeStatus.isNodeWritable() &&
+          (super.hasEnoughSpace(datanodeInfo, sizeRequired))) {
+        LOG.debug("Datanode {} is chosen. Required storage size is {} bytes",
             node.toString(), sizeRequired);
```
`node.toString()` is going to always be executed even with INFO level logs. This should be changed to just `node` and let the logger call toString on it internally if debug is enabled.

ChenSammi replied:
Good point.

sodonnel commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/NodeStateMap.java:
```diff
@@ -79,7 +86,10 @@ public void addNode(DatanodeDetails datanodeDetails, NodeStatus nodeStatus)
+      DatanodeInfo datanodeInfo = new DatanodeInfo(datanodeDetails, nodeStatus);
+      // Make sure nodes in topology tree has all DN info.
+      clusterMap.add(datanodeInfo);
+      nodeMap.put(id, datanodeInfo);
```
The node is already added to the cluster map on registration in `SCMNodeManager.register(...)`:

```
  @Override
  public RegisteredCommand register(
      DatanodeDetails datanodeDetails, NodeReportProto nodeReport,
      PipelineReportsProto pipelineReportsProto) {
      ...
        networkLocation = nodeResolve(dnsName);
        if (networkLocation != null) {
          datanodeDetails.setNetworkLocation(networkLocation);
        }

        clusterMap.add(datanodeDetails);
        nodeStateManager.addNode(datanodeDetails);
    ...
```

Why do we need to add it again here?

ChenSammi replied:
The add to topology action during SCMNodeManager.register(...) is removed in this patch.  

The goal is to make sure the node object added into topology is of DatanodeInfo type, so that when a node is chosen through topology API, this node will have the operation state information already, to save an extra call of NodeManager to get this information.

sodonnel replied:
@nandakumar131 @bshashikant What do you guys think about this change (putting the DatanodeInfo into the topology map)? I remember we talked about something like this with decommission, and concluded we were happier if the DatanodeDetails reference was only stored in a single place (nodeManager) and anything that needs it should request it from there. Our fear was that some future change could replace the DatanodeInfo instance stored in one place and not in the other.

sodonnel replied:
SCMContainerPlacementRackAware already has a NodeManager instance passed into it. I think this change would be cleaner and safer if we just used NodeManager to lookup the NodeStatus. Then all the locks are obeyed and DatanodeInfo is stored only in a single place.

The problem then is that the NetworkToplogy instance holds only a Node instance. But we need a DatanodeDetails instance (which extends Node) to query NodeManager to get the NodeStatus.

sodonnel commented on hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/node/states/TestNodeStateMap.java:
```diff
@@ -127,26 +129,26 @@ public void testGetNodeMethodsReturnCorrectCountsAndStates()
   @Test
   public void testConcurrency() throws Exception {
-    NodeStateMap nodeStateMap = new NodeStateMap();
+    //NodeStateMap nodeStateMap = new NodeStateMap();
```
This commented line should probably be removed, however I am not sure if any of these NodeStateMap changes are needed, as I am not sure if we need to add the node to the topology map in NodeStateMap, when it already happens in the register method.

ChenSammi replied:
Yes, this line should be removed.

sodonnel reviewed (changes requested):
These changes make sense to me, I have pointed out a couple of small things to check inline.

Another scenario for this problem is:

1. Replication Manager schedules a lot of replication commands on a host where they are queued.

2. The host is then decommissioned soon after.

In that case, all those inflight replications are still going to be executed.

Should we also check the DN state in the DN replication command handler, and if the DN is not IN_SERVICE, ignore them?

You can get the DN state inside `ReplicateContainerCommandHander` via:

```
context.getParent().getDatanodeDetails().getPersistedOpState()
```
So if we passed the context.getParent() into `new ReplicationTask(...)` it would be able to make that check when it is executed.

ChenSammi commented:
> 
> 
> These changes make sense to me, I have pointed out a couple of small things to check inline.
> 
> Another scenario for this problem is:
> 
>     1. Replication Manager schedules a lot of replication commands on a host where they are queued.
> 
>     2. The host is then decommissioned soon after.
> 
> 
> In that case, all those inflight replications are still going to be executed.
> 
> Should we also check the DN state in the DN replication command handler, and if the DN is not IN_SERVICE, ignore them?
> 
> You can get the DN state inside `ReplicateContainerCommandHander` via:
> 
> ```
> context.getParent().getDatanodeDetails().getPersistedOpState()
> ```
> 
> So if we passed the context.getParent() into `new ReplicationTask(...)` it would be able to make that check when it is executed.

Right. I have added the  DN state check in  ReplicateContainerCommandHandler#handle.

ChenSammi commented:
Thanks @sodonnel for the code review.  A new commit is uploaded to address the comments.

JacksonYao287 commented:
thanks @ChenSammi for this work! the changes looks good to me, but a little question!

if we make a change to 'clusterMap', should we do some more work about `SCMNodeManager# getClusterNetworkTopologyMap`, which will be called by other functions!

JacksonYao287 reviewed (changes requested):
```
  public NetworkTopology getClusterNetworkTopologyMap() {
    return clusterMap;
  }
```

sodonnel commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/ReplicateContainerCommandHandler.java:
```diff
@@ -60,13 +61,20 @@ public ReplicateContainerCommandHandler(
+          "for container {}", dn.getPersistedOpState(), containerID);
+      return;
+    }
+
```
I think this needs to go into the supervisor, so that when the task is executed by the executor, it checks to see the DN state and drops the command at that point. We could have a lot of replication commands queued which are past this point, and if the DN switches to a not IN_SERVICE state, ideally we want to prevent processing them all.

ChenSammi replied:
Good point.

sodonnel commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackAware.java:
```diff
@@ -275,28 +276,38 @@ private Node chooseNode(List<Node> excludedNodes, Node affinityNode,
+          .getNodeByUuid(datanodeDetails.getUuidString());
+      if (datanodeInfo == null) {
+        LOG.error("Failed to find the DatanodeInfo for datanode {}",
+            datanodeDetails);
```
You can get the nodeStatus directly via NodeManager:

```
  public NodeStatus getNodeStatus(DatanodeDetails datanodeDetails)
      throws NodeNotFoundException {
```

Its doing basically what you are doing here, but I think its safer as an immutable NodeStatus object is returned under a lock to you:

```
  public NodeStatus getNodeStatus(UUID uuid) throws NodeNotFoundException {
    lock.readLock().lock();
    try {
      DatanodeInfo dn = nodeMap.get(uuid);
      if (dn == null) {
        throw new NodeNotFoundException("Node not found in node map." +
            " UUID: " + uuid);
      }
      return dn.getNodeStatus();
    } finally {
      lock.readLock().unlock();
    }
  }
```

That means we don't need to worry about any side effects or concurrent modification here at all.

ChenSammi replied:
@sodonnel ,  it is really because the following super.hasEnoughSpace(datanodeInfo, sizeRequired)) statement needs a DatanodeInfo object as parameter so here getNodeByUuid is called.

sodonnel replied:
Ah, I missed the call to `hasEnoughSpace`. OK - I think this is good now.

ChenSammi commented:
> 
> 
> Thanks for addressing the review comments @ChenSammi. I think this change looks better now. I just have one more comment inline where you are getting the nodeStatus - I think you should call the existing NodeManager API to get it so it is all protected under a lock.
> 
> > Removing the dead dn from topology seems a good idea.
> 
> I think we should do that part under a separate Jira as its a slightly different change.

HDDS-5437 is created for track.

sodonnel reviewed (approved):
Latest changes LGTM, +1. 

@JacksonYao287 do you have any further comments / concerns before we commit?

JacksonYao287 commented:
this patch looks good to me , thanks @ChenSammi  for this work!
